### PR TITLE
nil: add meta.mainProgram

### PR DESCRIPTION
**Pull request by khaneliman**
## Description of changes
Add meta.mainPackage to remove deprecation warning. 
Related PR: https://github.com/NixOS/nixpkgs/pull/246386

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/language-servers/nil/default.nix [hunk 1]
```diff
@@ -26,11 +26,12 @@ rustPlatform.buildRustPackage rec {
...
-  meta = with lib; {
+  meta = {
     description = "Yet another language server for Nix";
     homepage = "https://github.com/oxalica/nil";
     changelog = "https://github.com/oxalica/nil/releases/tag/${version}";
-    license = with licenses; [ mit asl20 ];
-    maintainers = with maintainers; [ figsoda oxalica ];
+    license = with lib.licenses; [ mit asl20 ];
+    maintainers = with lib.maintainers; [ figsoda oxalica ];
+    mainPackage = "nil";
```
Comment (eclairevoyant) [inline pkgs/development/tools/language-servers/nil/default.nix:35]:
```suggestion
    mainProgram = "nil";
```

**Review by eclairevoyant — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/language-servers/nil/default.nix (hunk 1 above)
Comment (eclairevoyant) [reply pkgs/development/tools/language-servers/nil/default.nix:35]: unless i am missing something

Comment (khaneliman) [reply pkgs/development/tools/language-servers/nil/default.nix:35]: yikes... brain fail lol

**Comment by figsoda** [comment]
please revert the unrelated changes

**Review by Gerg-L — APPROVED** [review]

**Review by eclairevoyant — APPROVED** [review]

**Review by figsoda — APPROVED** [review]
I'll merge when ofborg passes

**Pull request merged**